heareval/predictions/task_predictions.py

```diff
@@ -904,6 +981,125 @@ def hparams_to_json(hparams):
+    test=fold01, val=fold02, train=fold03,04,01
+    ...
+    test=fold04, val=fold00, train=01..03
+    """
+    sorted_folds = tuple(sorted(folds))
+    assert len(sorted_folds) == len(set(sorted_folds)), "Folds are not unique"
+    num_folds = len(sorted_folds)
+    all_data_splits: List[Dict[str, List[str]]] = []
+    for fold_idx in range(num_folds):
+        test_fold = sorted_folds[fold_idx]
+        valid_fold = sorted_folds[(fold_idx + 1) % num_folds]
+        train_folds = list(set(sorted_folds) - {test_fold, valid_fold})
+        all_data_splits.append(
+            {
+                "train": train_folds,
+                "valid": [valid_fold],
+                "train+valid": train_folds + [valid_fold],
+                "test": [test_fold],
+            }
+        )
+        assert not set(train_folds).intersection(
+            {test_fold, valid_fold}
+        ), "Train folds are not distinct from the dev and the test folds"
+
+    return all_data_splits
+
+
+def aggregate_test_results(results: Dict[str, Dict[str, float]]) -> Dict[str, float]:
+    """
+    Aggregates test results over folds by mean and standard deviation
+    """
+    results_df = pd.DataFrame.from_dict(results, orient="index")
+    aggregate_results = {}
+    for score in results_df:
+        aggregate_results[f"{score}_mean"] = results_df[score].mean()
+        aggregate_results[f"{score}_std"] = results_df[score].std()
+
+    return aggregate_results
+
+
+def get_splits_from_metadata(metadata: Dict) -> List[Dict[str, List[str]]]:
+    """
+    Extracts the splits for training from the task metadata. If there are folds
+    present then this creates a set of k splits for each fold.
+    """
+    data_splits: List[Dict[str, List[str]]]
+    if metadata["split_mode"] == "trainvaltest":
+        # There are train/validation/test splits predefined. These are the only splits
+        # that will be considered during training and testing.
+        data_splits = [
+            {
+                "train": ["train"],
+                "valid": ["valid"],
+                "train+valid": ["train" + "valid"],
+                "test": ["test"],
+            }
+        ]
+    elif metadata["split_mode"] in ["new_split_kfold", "presplit_kfold"]:
+        folds = metadata["splits"]
+        # Folds should be a list of strings
+        assert all(isinstance(x, str) for x in folds)
+        # If we are using k-fold cross-validation then get a list of the
+        # splits to test over. This expects that k data folds were generated
+        # during pre-processing and the names of each of these folds is listed
+        # in the metadata["folds"] variable.
+        data_splits = data_splits_from_folds(folds)
+    else:
+        raise AssertionError(
+            f"Unknown split_mode: {metadata['split_mode']} in task metadata."
+        )
+
+    return data_splits
+
+
+def sort_grid_points(grid_point_results: List[GridPointResult]) -> None:
```

Comment:
Why are we sorting in-place? Why not return the sorted copy, and overwrite it when in the return? Nicer not to have side effects.